JP-3581: badpix_selfcal step

emolter opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3581](https://jira.stsci.edu/browse/JP-3581)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #[8369](https://github.com/spacetelescope/jwst/issues/8369)

<!-- describe the changes comprising this PR here -->
This PR adds a new optional step to the spec2pipeline that applies a median filter in the spectral direction of IFU data to flag warm and cool pixels.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

emolter commented:
@drlaw1558 @jemorrison This first draft of the step provides a "minimum working example" - it implements the median filtering in the spectral direction to find bad pixels, then flags them.  If an association is passed in, as is done in the regression test, then the flagging will occur on the pixelwise MIN of the background exposures, then get applied to both science and background exposures, which are saved into separate .fits files.  If a single _rate file is passed, then this will operate on that rate file, i.e., true self-calibration.

Let me know if the current behavior of the code is what is desired, and what needs to be changed.

@melanieclarke suggested to use the "other bad pixel" flag instead of simply "do not use"; should I go ahead and do that?

I ran into some issues with file I/O that I am unsure of how to resolve yet.  In short, if I pass an association into calwebb_spec2 (as opposed to just this step), then assign_wcs will pass an IFUImageModel into this step instead of a ModelContainer, meaning that the step never knows about the background exposures and reverts to operating on the science data alone.  If I run the step on an association, it's not obvious to me how to format the output such that it's easily usable downstream (filenames to use, how to save background exposures, how to update the association if at all).

Still marked as draft until those issues can be resolved, and I'd appreciate any suggestions you have.  

I'm working on adding a docs page now.

drlaw1558 commented:
Thanks @emolter ; that was super fast!

I think there would probably need to be some changes on the association-generator front to ensure that this step is getting passed an asn that includes all of the files that would be helpful.  If the pipeline notebooks that we're providing people with are modified to create them properly that would work, but it would be useful if the default spec2 association files downloaded from MAST also had the right information for this to work.  I.e., include all science, background, leakcal, etc files (probably not target acq though as those may be much shallower, noisier, and generally not like the science data) associated with a given observation in the spec2 asn file with exptype='selfcal'.  At the moment these (e.g., for MIRI MRS) often contain just a single 'science' member and nothing else.  That exptype should be ignored by all other spec2 steps, but then this self-calibration step can simply use all members of the asn list that aren't target_acq for use in the algorithm.

I don't think we want to explicitly apply the flagging to both science and background exposures here; at least not from a single asn file.  E.g., for PID 1523, Observation 1 is the science data, and Observation 2 the background.  The association files generated for Observation 2 are currently just a single background exposure, which has exptype=Science (since it's processed as science data in spec2 and gets used as a background in spec3).  As such, I can imagine file collisions if re-flagged data for the Obs2 exposures is written out both during Obs 1 and Obs 2 processing.  This should generally be the case I think when we care about applying the flagging to the background data too- when they get processed as Science data in spec2 so that they can be used as backgrounds in spec3.

We probably don't even need to explicitly ask the code to use only 'background' exposures if those exist, which would be simpler for the code.  By definition, the MIN algorithm should be able to handle being passed both science and background together as background should be fainter than source observations.  (If not, it suggests the backgrounds are bad and shouldn't be used alone anyway).  Essentially, things marked as exptype=Background aren't special in any way; flagging should just get applied to whatever the exposure is with exptype=Science in the lvl2 asn file, using everything in the asn file to generate the mask.  Does that make sense?

Setting other_bad_pixel in addition to DO_NOT_USE sounds fine to me.  It's a catch-all flag, and this is a pretty catch-all algorithm.

emolter commented:
Thanks for the detailed review @drlaw1558 !  Some quick responses point-by-point:

- Setting `other_bad_pixel` - sounds good
- Using all exposures in the asn to compute the min - sounds good
- Not flagging exposures marked as background - This was originally @jemorrison's suggestion.  I don't initially see how, as written, the file conflict you bring up would happen, because the spec2pipeline does not support marking more than one file in an association as science.  In the present code version, only a single output file is written representing the science exposure unless the optional `save_flagged_bkgd` parameter is True, and in that case a different file extension should be used (although I need to check this).  When the whole spec2pipeline is run, these flagged backgrounds are passed as a list into the `background` step.
- Updates to association rules: it sounds like you want an entirely new classification of exposures for the association jus for this step.  That sounds good in principle, although I wonder if adding all these extra files by default would confuse users. Perhaps it should be optional for association generation too, and turned off by default?
- PID 1523 - can you point me to an association file or files where one science exposure is used as a background for the other? I don't see any association files in the Box folder you linked to the ticket, and the one I was able to download for 1523 from MAST just has a single exposure in it

melanieclarke commented:
I'm also wondering about the effects of including all the files in the default association files in MAST.  I think the pipeline complains if a file is in the association but isn't on disk, even if it isn't used in the pipeline -- this already happens for TA files in spec3.  If everything that could be used for selfcal was listed, that might make it more difficult to use MAST association files for users not interested in doing selfcal.  I agree that some user support and documentation would be helpful though, because association files are not necessarily trivial to create.

melanieclarke reviewed (commented):
Some initial comments/suggestions here.  I can test with real data and review again when things are settled.

drlaw1558 commented:
As a slightly more normal science example (PID 1523 is a somewhat unusual mosaic), see PID 1328 Observations 15 and 16.  Observation 15 is the science exposures on a galaxy target, and Observation 16 is the dedicated background exposures.  Both Observation 15 and Observation 16 get processed through spec2 as 'science' type data, and it's only in spec3 that Observation 16 is treated as a background for purposes of the master_background step.  As such, having Observation 15 write out flagged backgrounds is unnecessary- it should happen when Observation 16 is processed.

asns are certainly the trick though; I agree changing *everything* to add a bunch more entries to an association is potentially risky.  At the same time, *not* doing so makes it harder for users to take advantage of the step...

Maybe for the time being we should take the safe route; rely on users creating a spec2 association that can take advantage of the code.  I can certainly add that into the MRS notebook that we provide to the community, and add some documentation to JDox about how to do it more generally.  If the selfcal step is passed an association with just one member (or simply passed a file instead of an association), it should print an error message and skip the step.  There are a lot of failure modes using just a single exposure (emission lines in particular), and I can see users trying to turn it on without setting it up properly getting caught by that.  We could have some parameter that says "I know what I'm doing, really really run on just a single input image" to enable it.

jemorrison reviewed (commented):
I don't have much to add on the code changes. Melanie did a thorough job on that front. Personally I would hold off on association rule changes for now.  That could have many unintended consequences that could led to errors. If you can provide documentation and maybe a notebook to users on how to construct an association to run this option step that might be a first step. Then after we get feed back from folks we can see if the community would want this easier to run and think then about an association rules change (a general association rules change to make this step run  frankly scares me a bit). When I was using David's notebook to flag warm pixels I constructed by own associations containing the backgrounds. It is fairly easy to do.

emolter commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -246,6 +248,15 @@ def process_exposure_product(
 
+        # self-calibrate bad/warm pixels, and apply to both background and science
+        # skipped by default for all modes
+        result = self.badpix_selfcal(calibrated, members_by_type['background'])
```
I believe this can remain the way it is for now, given the current state of that conversation, but let me know if you think otherwise

melanieclarke replied:
I think this still needs updating - the step will need to get 'selfcal' and 'background' members from the association file, not just background.

emolter commented:
I just pushed a new version that attempts to respond to all the comments.  The changes are:

- set DQ flag to "DO_NOT_USE" + "OTHER_BAD_PIXEL"
- removed option to save flagged background exposures to file
- added warning and step skip if a single science exposure is passed in, added `force_single` parameter to override this
- step now uses science and background exposures identically when creating the MIN exposure
- added mention of the step to calwebb_spec2 pipeline docs

Some additional things that I feel are still up in the air:

- I retained the behavior that background exposures are flagged when passed through the entire calwebb_spec2 pipeline; the effect of this is that the background_sub step would see background exposures with the flags applied.  It seemed to me like that is a good thing; however, if the notebook examples we create are re-using exptype="background" for e.g. leakcal files, then that might cause problems
- That notebook still needs to be created.  IMO it might be best for either @melanieclarke or @drlaw1558 to make a draft of this, or at least an example of what you would like the input user-created asn files to look like; then I can test that through both the individual step and the whole calwebb_spec2 pipeline to ensure that it behaves as intended.
- A regression test is still needed.  I'm holding off on this until everything else is finalized, simply because it's tedious to continually modify the inputs and truth files that are uploaded to the artifactory.

emolter commented:
After some discussion with @melanieclarke I went ahead and assumed an example association file structure.  You can see that assumption in the testing suite [here](https://github.com/spacetelescope/jwst/pull/8500/files#diff-64d40d97abf3f616620dbf14c3af9b126d545d44d00e7e279b3de9c362b9def0R125).

With the current setup, all files in the association ("science", "background", and "selfcal") are used to make the MIN background.  Flags are applied to anything marked "science" or "background", and these are returned to `calwebb_spec2.py` such that the flagged backgrounds are passed along into subsequent steps like `bkg_subtract`.  The "selfcal" files are not passed along in any way.

If `save_results=True`, then the flagged science and background exposures are each saved to separate filenames; any files marked "selfcal" are not saved.

Let me know if these are the desired behaviors and what else needs to be changed, and then I will update the docs accordingly.  I can also make a very basic example notebook if you would like.

emolter commented:
@drlaw1558 it was suggested during our planning meeting today to tag you again on this for feedback - the step is "done" as far as I'm concerned, but it would be helpful if you could verify whether the asn looks like you want it to look, and also test this with real data (as well as any other feedback you have).  I have tested it locally with some data Jane gave me, simply setting some of the "background" exposures to "selfcal", but I don't have any data locally yet where the "selfcal" data is distinct from the backgrounds.

drlaw1558 commented:
Thanks @emolter ; I'll try to take a look later this afternoon if possible.

drlaw1558 commented:
Hi @emolter ; just had a chance to start trying out this code.  A few things that I noticed:

- The smoothing is happening in the wrong direction; the code will need to query the dispersion axis (DISPAXIS, or model.meta.wcsinfo.dispersion_direction) to figure out in which dimension to apply the smoothing.  '1' would be X, and '2' would be Y.  I think if the keyword is undefined (imaging) then it should use a 2d kernel.

- The badpix_selfcal_step code is currently written distinguishing between 'science', 'background', and 'selfcal'.  At the moment providing 'selfcal' inputs does nothing, they just get ignored and it uses the single science frame.  It looks like providing 'background' inputs via an association file actually populates bkg_list as if it were user-defined, bypassing both 'if' (ModelContainer) and 'elif' (IFUImageModel) clauses in the step.py code.  This doesn't happen if provided 'selfcal' inputs instead; the ModelContainer code gets skipped and the IFUImageModel code doesn't populate selfcal info into the list.

- To avoid confusion later, it might be helpful to call the variable 'bkgd' in the step.py and 'medbg' in the selfcal.py codes something like 'minimg' instead, since it's neither a median nor necessarily made from background exposures.

emolter commented:
> Hi @emolter ; just had a chance to start trying out this code. A few things that I noticed:
> 
> * The smoothing is happening in the wrong direction; the code will need to query the dispersion axis (DISPAXIS, or model.meta.wcsinfo.dispersion_direction) to figure out in which dimension to apply the smoothing.  '1' would be X, and '2' would be Y.  I think if the keyword is undefined (imaging) then it should use a 2d kernel.
> * The badpix_selfcal_step code is currently written distinguishing between 'science', 'background', and 'selfcal'.  At the moment providing 'selfcal' inputs does nothing, they just get ignored and it uses the single science frame.  It looks like providing 'background' inputs via an association file actually populates bkg_list as if it were user-defined, bypassing both 'if' (ModelContainer) and 'elif' (IFUImageModel) clauses in the step.py code.  This doesn't happen if provided 'selfcal' inputs instead; the ModelContainer code gets skipped and the IFUImageModel code doesn't populate selfcal info into the list.
> * To avoid confusion later, it might be helpful to call the variable 'bkgd' in the step.py and 'medbg' in the selfcal.py codes something like 'minimg' instead, since it's neither a median nor necessarily made from background exposures.

Hi @drlaw1558 thanks for testing the step.  A clarification question RE the second point: how exactly were you calling the step? (e.g., through the whole pipeline, step.from_cmdline(), badpix_selfcal.call(), or something else?)

The other two comments are clear.  I'll push a fixed version, hopefully be the end of the day

emolter commented:
![directional-flagging-test](https://github.com/spacetelescope/jwst/assets/21211014/469fc37e-b94a-4df1-9c21-44bd7c146fdf)

Here's a piece of test "data" that I am now generating in the unit testing suite.  The colors represent relative brightness, with black being dark; the vertical stripes are meant to represent a spectral trace.  You can see the testing code [here](https://github.com/spacetelescope/jwst/pull/8500/files#diff-64d40d97abf3f616620dbf14c3af9b126d545d44d00e7e279b3de9c362b9def0R265).  To be sure I am getting the direction right, in this case we would expect all four outlier points to be flagged if DISPAXIS=2, is that correct?  And if DISPAXIS=1 or None, we expect only one outlier, the point that sits atop one of the positive stripes and is therefore brighter than the others, to be flagged?

drlaw1558 commented:
@emolter In response to your question above, I was running spec2.call() from a notebook, with spec2dict['badpix_selfcal']['skip']=False

drlaw1558 commented:
The new dispaxis selector code looks right to me, and I agree with you on expected results from that test.

melanieclarke commented on docs/jwst/badpix_selfcal/arguments.rst:
```diff
@@ -0,0 +1,14 @@
+
+``--kernel_size`` (integer, default=15)
+  The size of the kernel to use for the median filter, which is applied 
+  in the spectral (Y-axis) direction to make the smoothed image.
```
```suggestion
  in the spectral direction to make the smoothed image.
```

melanieclarke commented on docs/jwst/badpix_selfcal/description.rst:
```diff
@@ -0,0 +1,52 @@
+--------
+The ``badpix_selfcal`` step flags bad pixels in the input data using a self-calibration 
+technique based on median filtering along the spectral axis. 
+When background exposures are available, those are used in combination with the science
+exposure to identify bad pixels; when unavailable, the step will be skipped with a warning
+unless the ``force_single`` parameter is set True. In that case, the science data alone is
+used as its own "background".
```
Needs updating, since it might not be just background exposures. Maybe "additional exposures"?

melanieclarke commented on docs/jwst/badpix_selfcal/description.rst:
```diff
@@ -0,0 +1,52 @@
+-------------
+The input data must be in the form of a `~jwst.datamodels.IFUImageModel` or 
+a `~jwst.datamodels.ModelContainer` containing exactly one
+science exposure and any number of background exposures. A fits or association file 
```
Here too - 
```suggestion
science exposure and any number of additional exposures. A fits or association file 
```

melanieclarke commented on docs/jwst/badpix_selfcal/description.rst:
```diff
@@ -0,0 +1,52 @@
+The algorithm relies on the assumption that bad pixels are outliers in the data along
+the spectral axis. The algorithm proceeds as follows:
+
+* A combined background image is created. If background exposures are available, 
+  the pixelwise minimum of all background and science exposures is taken. 
+  If no background exposures are available, the science data itself is passed in 
+  without modification, serving as the "background image" for the rest of the procedure, 
+  i.e., true self-calibration.
```
Here too

melanieclarke commented on docs/jwst/badpix_selfcal/description.rst:
```diff
@@ -0,0 +1,52 @@
+  If no background exposures are available, the science data itself is passed in 
+  without modification, serving as the "background image" for the rest of the procedure, 
+  i.e., true self-calibration.
+* The combined background image is median-filtered, ignoring NaNs, along the spectral (Y-) axis 
```
```suggestion
* The combined background image is median-filtered, ignoring NaNs, along the spectral axis 
```

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,88 @@
+        Flagged data model
+    """
+
+    input_model.dq[flagged_indices] = pixel["DO_NOT_USE"] + pixel["OTHER_BAD_PIXEL"]
```
Here, it's probably advisable to use a bit-wise or ('|') instead of just setting directly to these two flags. That way, you can preserve any additional informational flags that pixel might have already had.

emolter replied:
I updated it to be a `+=` operation instead.  My understanding of the flagging scheme is that this will preserve all the information because all sums of DQ flags are unique.  Is that right?  It wasn't immediately obvious to me why using `or` is correct here.

melanieclarke replied:
It should be |, not +.  You need a bitwise-or because a pixel might already have one of the flags you're trying to set.  If it already has DO_NOT_USE, then += DO_NOT_USE would set it to 2 instead of 1, which is a completely different flag.

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+    All input exposures in the association file (or manually-provided bkg_list) are combined
+    into a single background model using a MIN operation. The bad pixels are then identified
+    using a median filter and percentile cutoffs, and applied to the science data by setting
+    the flagged pixels to NaN and the DQ flag to DO_NOT_USE + OTHER_BAD_PIXEL.
```
Maybe mention that errors and variances are also set to NaN

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No background exposures provided for self-calibration. Skipping step.")
```
```suggestion
            log.warning("No additional exposures provided for self-calibration. Skipping step.")
```

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No background exposures provided for self-calibration. Skipping step.")
+            self.record_step_status(input_sci, "badpix_selfcal", success=False)
```
This function will currently set skip=True for the step. It's probably better to just directly set the calibration step status to SKIPPED.

emolter replied:
How does the change you are proposing interact with [this ticket](https://jira.stsci.edu/browse/JP-3631)?  And, related, what is the point of having a method called `record_step_status()` if it should not be used?

melanieclarke replied:
That ticket is why I mentioned it.  It's preferable not to set step.skip = True from inside a pipeline step, because it can have unintended consequences.  It's fine to call record_step_status with success=True, but until we get it updated, best not to call it with success=False.  Or, send in a fix for the function and call it in both places here!

emolter replied:
You're saying the only way skipping a step should happen is if it's specified in the inputs or by a reference file, is that right?  In that case, I assume I should just make this throw an exception?

melanieclarke replied:
Setting the skip attribute in the step should only happen at the pipeline level, if the step is truly inappropriate to run on the input data (e.g. an instrument mode mismatch), or else if the user directly specified it.  

It's still okay to skip a step when some internal condition is encountered!  You can just warn, set the SKIPPED status, and return.  You just don't want to set it in the step attribute, because that would still skip the step if some other input data was passed, even if it might otherwise meet your internal criteria.

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+        # all exposures are combined into a single background model using a MIN operation.
+        selfcal_list = [input_sci] + [dm.open(k) for k in selfcal_list]
+
+        # collapse background dithers into a single background model
```
```suggestion
        # collapse all dithers into a single background model
```

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+            for i, background_model in enumerate(bkg_list_asn):
+                bkg_list_asn[i] = badpix_selfcal.apply_flags(background_model, bad_indices)
+
+        self.record_step_status(input_sci, "badpix_selfcal", success=True)
```
Again, safer to just set the calibration step status

melanieclarke commented on jwst/master_background/master_background_step.py:
```diff
@@ -236,29 +236,48 @@ def copy_background_to_surf_bright(spectrum):
 
-def split_container(container):
-    """Divide a ModelContainer with science and background into one of each
+def split_container(container: ModelContainer, fields=['science', 'background']):
```
Should we move this function to a library somewhere, since it's now needed in multiple places?

emolter replied:
I am all for that, but I wonder where it should go? How about with ModelContainer inside datamodels.container?

melanieclarke replied:
That sounds appropriate to me.

braingram replied:
A few words of caution here:
- split_container is also used in [calwebb_spec3](https://github.com/spacetelescope/jwst/blob/d6435f18465155fa32fecb8c4a94ab192dcec1d0/jwst/pipeline/calwebb_spec3.py#L12)
- the current function only works for a `ModelContainer` read from an association (or a "derivative" of one). A container made with a list of models won't have for instance `asn_pool_name`
- the current function attempts to keep the "member" list updated (for the "science" container) but won't produce expected results if a models `meta.asn.exptype` is changed (the split happens based on the model metadata but the member split happens based on the association values)

If I'm following the changes in this PR, only the first science model is used (so the science container is discarded after it's split). This discards the association information and in this case a list of models grouped by `meta.asn.exptype` seems easier to follow (and less prone to the issues above) vs using `split_container`.

emolter replied:
@braingram ok I hear you RE the `split_container` function only operating on a ModelContainer that is read from an association.  But it also feels a bit strange to define a new function in this step that, well, splits a container, even if the two operate slightly differently.  I guess philosophically I would expect a function to exist that will split a container by exposure type, and have it work on any valid ModelContainer object.  So I guess for now, because I know ModelContainer is soon being deprecated anyway, I'll write the separate function, but something to keep in mind perhaps for ModelLibrary.

With regards to being used also in calwebb_spec3, yes, that's true, and the proposed changes are fully back-compatible with the old behavior.  But that point is moot if I make your suggested fix and bypass split_container entirely.

melanieclarke commented on jwst/regtest/test_miri_mrs_badpix_selfcal.py:
```diff
@@ -0,0 +1,28 @@
+
+    rtdata = rtdata_module
+    rtdata.get_asn("miri/mrs/jw01204-o021_20240127t024203_spec2_00010_asn.json")
+    BadpixSelfcalStep.call(rtdata.input, skip=False, save_results=True, flagfrac=0.0005)
```
Might also be a good idea to add a regtest for running it from the spec2 pipeline, with skip=False.

melanieclarke reviewed (commented):
I haven't tried running this yet, but I have few quick comments just from code inspection first.

I'll make myself a test case and give it a go soon!

melanieclarke commented:
Testing now by running the spec2 pipeline on a NIRSpec IFU data set, setting skip=False for the badpix_selfcal step.  I defined an association file with one science and 3 selfcal members.

Changes are definitely needed in spec2: running this way, the selfcal files are ignored and the step bails with a warning message saying it only has one image to work with.

Also, I think because there are no background images, the return value from the step causes the pipeline to crash in a later step:
```
RuntimeError: Traceback (most recent call last):
  File "/Users/mclarke/scsb/repository/other_jwst/nm_jwst/jwst/pipeline/calwebb_spec2.py", line 130, in process
    result = self.process_exposure_product(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mclarke/scsb/repository/other_jwst/nm_jwst/jwst/pipeline/calwebb_spec2.py", line 273, in process_exposure_product
    members_by_type['background'][i] = bkg_imprint_sub
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
TypeError: 'tuple' object does not support item assignment
```

emolter commented:
ok, thanks, I will work on it

melanieclarke commented:
Coercing the code a little bit to get it to run on my test case, it looks like this step does help quite a bit to clean up residual artifacts.  It's not perfect with default parameters, but it's pretty good!

I think it might be useful to separate out the upper and lower thresholds for bad pixels in the input parameters.  If the flag fraction needs to be tuned, you might need to access both ends of the distribution separately.

braingram commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,153 @@
+                            and selfcal exposure types in the association file")
+
+            # in calwebb_spec2 there should be only a single science exposure in an association
+            input_sci = sci_models[0]
```
Perhaps a warning (or error) if `sci_models` contains >1 model?

drlaw1558 commented:
@hbushouse I think we determined that it was safer for now to rely on users (or our provided pipeline notebooks) to generate the necessary ASN files, rather than risk unintended consequences for DMS this late in a build.  That should be enough to allow folks to start testing the step further, and we can always add something official later.

emolter commented:
> @hbushouse I think we determined that it was safer for now to rely on users (or our provided pipeline notebooks) to generate the necessary ASN files, rather than risk unintended consequences for DMS this late in a build. That should be enough to allow folks to start testing the step further, and we can always add something official later.

Agreed, this is what I understood to be the consensus.

emolter commented:
Something appears to be very wrong with the CI tests, since it can't even find the changelog entry, but I'm assuming that has nothing to do with this PR in particular.  In any case I started a [regression test run here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1517/).

Current code version attempts to respond to the obvious issue brought up by @melanieclarke and @drlaw1558 that the step wouldn't run as part of the pipeline.  I ran Melanie's test case and it went through fine, and to my untrained eye the output _cal file looked reasonable.  I also added a regression test where calwebb_spec2 is run on two nearly identical association files and the output of badpix_selfcal is checked; the only difference between the associations is that in one of them, the background exposures have exptype="background" and in the other they have exptype="selfcal"

I also tried to respond to all the other comments.  Those that I was not 100% sure about I left as unresolved with a comment; let me know if you like the solution I went with or not.

Sorry the last version was a bit of a mess.  I hope this one is at least a bit improved.

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,188 @@
+    """
+    selfcal_list_user = selfcal_list
+
+    if isinstance(input, tuple):
+        input_sci = dm.open(input[0])
+        selfcal_list = list(input[1]) + list(input[2])
+        bkg_list_asn = input[1]
```
This works with the way it is called in spec2 now, but it is a little convoluted.  Why not just have the step accept a selfcal_list, used only for selfcal, and a bkg_list with background files to update?  That way a user directly calling the step could also distinguish between selfcal files and background files, without having to make an association file, or know that they should put them in a tuple in the input variable.

emolter replied:
this is a great suggestion.  In working on this, I also realized it might be more intuitive for users if, in the case that an .asn file with selfcal exposures AND selfcal_list is not None, the selfcal_list were appended to the association selfcal list (rather than having the user input override it).  What do you think of that?

melanieclarke replied:
That sounds useful to me!

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,188 @@
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No selfcal or background exposures provided for self-calibration. \
+                        Skipping step. If you want to force self-calibration with the science \
+                        exposure alone (generally not recommended), set force_single=True.")
```
These warnings are not printing the way I think you intended - they show up with lots of spaces where the \ appears.  This should work better:
```suggestion
            log.warning("No selfcal or background exposures provided for self-calibration. "
                        "Skipping step. If you want to force self-calibration with the science "
                        "exposure alone (generally not recommended), set force_single=True.")
```

The rest of the warnings in this module should be similarly updated.

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,188 @@
+    skip = boolean(default=True)
+    """
+
+    def process(self, input, selfcal_list=[]):
```
In Python, it's preferable not to have a mutable value as a default for a parameter.  It's better to set the default to None, then catch None as input and change it to a list inside the code, like:

```
if selfcal_list is None:
    selfcal_list = []
```

melanieclarke commented on docs/jwst/badpix_selfcal/arguments.rst:
```diff
@@ -0,0 +1,14 @@
+==============
+The ``badpix_selfcal`` step has the following optional arguments.
+
+``--flagfrac`` (float, default=0.001)
+  The fraction of pixels to flag as outliers on each of the low-flux and high-flux
+  sides of the smoothed-subtracted image.
```
Docs need updating for the new flagfrac_lower and flagfrac_upper parameters.  (Thanks for those!)

emolter commented:
Again not sure why the CI tests are acting up, but starting a [new regtest run here.](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1522/)

drlaw1558 commented:
Looking pretty good to me and giving the results I'd expect in the cases I've tested it on.  Attached is an example of running it on MRS point source data with no true dedicated background.  I've dialed the flagging to 1% to be more obvious, and you can see it finding and (with the aid of pixel_replace) fixing bright blotches that weren't known about before.  Also tested it out on some NIRSpec IFU dithered point source observations, and looks like it's working well there too.

<img width="1265" alt="selfcal" src="https://github.com/spacetelescope/jwst/assets/11397442/7a5f9b36-3304-4cb9-b6d6-cce1d182ed27">

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,93 @@
+    elif dispaxis == 1:
+        kern_size = (1, kernel_size)
+
+    smoothed = medfilt(minimg, kern_size)
```
This line sometimes raises `RuntimeWarning: All-NaN slice encountered` for NaNs in the data, but that's a normal condition for NIRSpec IFU data at least.  Can we please catch and filter the warning?

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,194 @@
+        selfcal_3d = []
+        for i, selfcal_model in enumerate(selfcal_list):
+            selfcal_3d.append(selfcal_model.data)
+        minimg = np.nanmin(np.asarray(selfcal_3d), axis=0)
```
This line also raises a RuntimeWarning for NaNs that should be filtered.

melanieclarke reviewed (commented):
I think this is looking good now - just a few tiny things left.

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,194 @@
+        if (len(selfcal_list + bkg_list) == 0) and (not self.force_single):
+            log.warning("No selfcal or background exposures provided for self-calibration."
+                        "Skipping step. If you want to force self-calibration with the science"
+                        "exposure alone (generally not recommended), set force_single=True.")
```
Already updated!

melanieclarke commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,194 @@
+            dispaxis = input_sci.meta.wcsinfo.dispersion_direction
+        except AttributeError:
+            log.warning("Dispersion axis not found in input science image metadata."
+                        "Kernel for self-calibration will be two-dimensional.")
```
Already updated!

emolter commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,93 @@
+        Flagged data model
+    """
+
+    input_model.dq[flagged_indices] += pixel["DO_NOT_USE"] + pixel["OTHER_BAD_PIXEL"]
```
I'm still looking for clarification about this @melanieclarke : my understanding of the DQ flag scheme was that the values were all factor-of-two different, such that encoding multiple flags was allowed by simply adding the flag values.  If this is the case, then why is the OR operator desired here?

emolter replied:
sorry, didn't see you responded already

melanieclarke replied:
That's correct, but they are not factor of 2 different from themselves!  The or is needed in case the pixel already has one of the flags you are trying to set.  This is probably unlikely for DO_NOT_USE in this case, since they wouldn't be flagged as bad if they were already NaN, but it's very possible that a pixel might already have an OTHER_BAD_PIXEL flag (some existing NIRSpec darks use that value).  In that case, adding here would set the pixel to 2 * OTHER_BAD_PIXEL, which is a completely different flag.

melanieclarke replied:
And I missed your reply while I was typing in a new one. :)

melanieclarke reviewed (approved):
Everything looks good now, thanks for all the updates!

braingram commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,197 @@
+    bkg_suffix = "badpix_selfcal"
+
+    spec = """
+    flagfrac_lower = float(default=0.001)  #fraction of pixels to flag on the low-flux end
```
There are some range checks in `badpix_selfcal` that could be represented in this spec by setting `min` and `max`. Here's an example from resample:
https://github.com/spacetelescope/jwst/blob/master/jwst/resample/resample_step.py#L43
Would you update the spec to use whatever relevant constraints can be represented?

braingram commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,197 @@
+    """
+
+    class_alias = "badpix_selfcal"
+    bkg_suffix = "badpix_selfcal"
```
Where is `bkg_suffix` used?

braingram commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,197 @@
+
+import logging
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
```
I'm honestly not that familiar with the pipeline logging stuff but is this needed at the import level?

I think within the `Step` `self.log` can be used (it's set up [here](https://github.com/spacetelescope/stpipe/blob/d1a0a6d25de45400e315afa4fa3cee8633a6b545/src/stpipe/step.py#L366) in stpipe and appears to be used by other steps).

Perhaps @hbushouse can advise on which should be used.

emolter replied:
using `self.log` does seem to work identically, so going with @braingram suggestion

braingram reviewed (commented):
I'm going to leave this as a comment since it's mostly questions, comments and a few suggestions.

emolter commented on jwst/regtest/test_miri_mrs_badpix_selfcal.py:
```diff
@@ -0,0 +1,57 @@
+        fname = f"jw01204021001_02101_00004_mirifulong_badpix_selfcal_bkg_{idx}.fits"
+        if run_pipeline == "run_pipeline_background":
+            assert os.path.isfile(fname)
+            os.remove(fname)
```
I added a FITSDiff for the flagged background files.  However, with the current test setup, the `os.remove` is still required.  My understanding is that it's related to [this discussion](https://github.com/pytest-dev/pytest/issues/9151).

emolter replied:
Because they are identical and have the same filename, the file gets overwritten and the test passes.

emolter replied:
But yeah, I suppose it might be better to just split this into two tests instead of trying to be tricky with parametrize

braingram replied:
Oh I think I'm starting to understand. Yeah splitting these (if they're using the same directory) sounds good.

Also I tried to run this test locally and failed to find `jwst-pipeline/dev/miri/mrs/jw01204021001_02101_00004_mirifulong_rate.fits` (listed in the association) on artifactory. Is it still being uploaded?

emolter replied:
@braingram try now? I accidentally uploaded to the wrong folder

braingram replied:
Thanks! I was able to pull down the tests and successfully run the `test_miri_mrs_badpix_selfcal`.

emolter commented:
started a new [regtest run here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1528/).  Hopefully I got all the artifactory uploads in place.

emolter commented:
new regtest run with all suggested changes [is here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1529/)

braingram commented on docs/jwst/badpix_selfcal/arguments.rst:
```diff
@@ -0,0 +1,18 @@
+  The size of the kernel to use for the median filter, which is applied 
+  in the spectral direction to make the smoothed image.
+
+``--save_flagged_bkgd`` (boolean, default=False)
```
```suggestion
``--save_flagged_bkg`` (boolean, default=False)
```

emolter commented:
Latest regtest result is clean

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,91 @@
+        Flagged data model
+    """
+
+    input_model.dq[flagged_indices] |= pixel["DO_NOT_USE"] + pixel["OTHER_BAD_PIXEL"]
```
Does this really work correctly, in terms of handling our bit-encoded DQ flags? I would think that the generic "|" operator would only care whether any of the values involved are non-zero. Don't you need to use `np.bitwise_or()` here for the desired result? Otherwise you may be wiping out the bit-encoded values in `input_model.dq` and simply resetting them to either zero or one.

melanieclarke replied:
"|" is a bitwise operator in Python ("or" is the logical or operator).  In Python, `1 | 2 == 3`.

hbushouse replied:
Whoops, thanks for the clarification. Learn something new every day!  ;-)

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+    class_alias = "badpix_selfcal"
+
+    spec = """
+    flagfrac_lower = float(default=0.001, min=0.0, max=0.5)  #fraction of pixels to flag on the low-flux end
+    flagfrac_upper = float(default=0.001, min=0.0, max=0.5)  #fraction of pixels to flag on the high-flux end
+    kernel_size = integer(default=15, min=1)  #size of kernel for median filter
+    force_single = boolean(default=False)  #force single input exposure
+    save_flagged_bkg = boolean(default=False)  #save flagged background exposures to file
```
```suggestion
    flagfrac_lower = float(default=0.001, min=0.0, max=0.5)  # fraction of pixels to flag on the low-flux end
    flagfrac_upper = float(default=0.001, min=0.0, max=0.5)  # fraction of pixels to flag on the high-flux end
    kernel_size = integer(default=15, min=1)  # size of kernel for median filter
    force_single = boolean(default=False)  # force single input exposure
    save_flagged_bkg = boolean(default=False)  # save flagged background exposures to file
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,91 @@
+
+    Parameters
+    ----------
+    minimg: np.ndarray
+        Selfcal data of shape (x, y), i.e., after some operation has
+        already been taken to combine multiple exposures,
+        typically a MIN operation.
+    flagfrac_lower: float
+        Fraction of pixels to flag on the low end
+    flagfrac_upper: float
+        Fraction of pixels to flag on the high end
+    kernel_size: int
+        Size of kernel for median filter
+    dispaxis: int
```
```suggestion
    minimg : np.ndarray
        Selfcal data of shape (x, y), i.e., after some operation has
        already been taken to combine multiple exposures,
        typically a MIN operation.
    flagfrac_lower : float
        Fraction of pixels to flag on the low end
    flagfrac_upper : float
        Fraction of pixels to flag on the high end
    kernel_size : int
        Size of kernel for median filter
    dispaxis : int
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,91 @@
+
+    Returns
+    -------
+    flagged_indices: np.ndarray
```
```suggestion
    flagged_indices : np.ndarray
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,91 @@
+
+    Parameters
+    ----------
+    input_model: dm.IFUImageModel
+        Input science data to be corrected
+    flagged_indices: np.ndarray
```
```suggestion
    input_model : dm.IFUImageModel
        Input science data to be corrected
    flagged_indices : np.ndarray
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal.py:
```diff
@@ -0,0 +1,91 @@
+
+    Returns
+    -------
+    output_model: dm.IFUImageModel
```
```suggestion
    output_model : dm.IFUImageModel
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+
+
```
```suggestion
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+
+        Parameters
+        ----------
+        input: JWST data model or association
+            input science data to be corrected, or tuple of (sci, bkg, selfcal)
+
+        selfcal_list: list of ImageModels or filenames, default None
+            exposures to include as part of median background model used to find bad pixels,
+            but that are not flagged and returned as background exposures
+
+        bkg_list: list of ImageModels or filenames, default None
+            exposures to include as part of median background model used to find bad pixels,
+            and that are flagged and returned as background exposures
+
+        Returns
+        -------
+        output: JWST data model or association
```
```suggestion
        input : JWST data model or association
            input science data to be corrected, or tuple of (sci, bkg, selfcal)

        selfcal_list : list of ImageModels or filenames, default None
            exposures to include as part of median background model used to find bad pixels,
            but that are not flagged and returned as background exposures

        bkg_list : list of ImageModels or filenames, default None
            exposures to include as part of median background model used to find bad pixels,
            and that are flagged and returned as background exposures

        Returns
        -------
        output : JWST data model or association
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list + bkg_list) == 0) and (not self.force_single):
+            self.log.warning("No selfcal or background exposures provided for self-calibration. "
+                        "Skipping step. If you want to force self-calibration with the science "
+                        "exposure alone (generally not recommended), set force_single=True.")
```
This should probably be broken up into 2 or 3 separate message lines, rather than one big long one.

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+        if self.save_flagged_bkg:
+            self.save_bkg(bkg_list)
+
+        input_sci.meta.cal_step.badpix_selfcal = 'COMPLETE'
```
Has an update been made to stdatamodels core schema to define the new `cal_step.badpix_selfcal` attribute? At some point in the future we'll also need to file a JWSTKD ticket to add the new keyword to the official dictionary. At this point it's not critical, because the step is skipped by default and hence the keyword won't be used in products coming out of operations.

hbushouse replied:
You should go ahead and submit a PR against stdatamodels to at least get it added into our schema. I'll give myself the action to file a JWSTKD ticket to have it added to the jwst keyword dictionary.

emolter replied:
Sounds good.  That [PR is here](https://github.com/spacetelescope/stdatamodels/pull/305/files).

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+
+    Parameters
+    ----------
+    input: str
+        input science exposure or association
+
+    selfcal_list: list of ImageModels or filenames, default None
+        exposures to include as part of median background model used to find bad pixels,
+        but that are not flagged and returned as background exposures
+
+    bkg_list: list of ImageModels or filenames, default None
+        exposures to include as part of median background model used to find bad pixels,
+        and that are flagged and returned as background exposures
+
+    Returns
+    -------
+    input: JWST data model or association
```
```suggestion
    input : str
        input science exposure or association

    selfcal_list : list of ImageModels or filenames, default None
        exposures to include as part of median background model used to find bad pixels,
        but that are not flagged and returned as background exposures

    bkg_list : list of ImageModels or filenames, default None
        exposures to include as part of median background model used to find bad pixels,
        and that are flagged and returned as background exposures

    Returns
    -------
    input : JWST data model or association
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+    input: JWST data model or association
+        input science data to be corrected
+
+    selfcal_list: list of ImageModels or filenames to use for selfcal
```
```suggestion
    selfcal_list : list of ImageModels or filenames to use for selfcal
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+
+    Parameters
+    ----------
+    container: ModelContainer
```
```suggestion
    container : ModelContainer
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+    container: ModelContainer
+        input data
+
+    exptype: str
```
```suggestion
    exptype : str
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+
+    Returns
+    -------
+    split_list: list of lists
```
```suggestion
    split_list : list of lists
```

hbushouse commented on jwst/badpix_selfcal/badpix_selfcal_step.py:
```diff
@@ -0,0 +1,210 @@
+    return input_sci, selfcal_list, bkg_list
+
+
+def split_container_by_asn_exptype(container: dm.ModelContainer, exptypes: list) -> list:
```
This might be useful as a general utility function, perhaps over in `lib` somewhere.

emolter replied:
See conversation with @braingram above, probably marked as resolved now.  We discussed this possibility and decided there was no point in doing anything general because this will be replaced with ModelLibrary at some point in the future.  I have already asked for the ability to do this as part of what eventually gets shipped with the implementation of ModelLibrary

braingram replied:
See: https://github.com/spacetelescope/jwst/pull/8500#discussion_r1633685815

hbushouse replied:
You guys are way ahead of me!

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -243,6 +245,18 @@ def process_exposure_product(
                         raise RuntimeError('Cannot determine WCS.')
 
         # Steps whose order is the same for all types of input:
+        # self-calibrate bad/warm pixels, and apply to both background and science
```
Minor nit, but to the uninitiated the statement that you're "self calibrating bad/warm pixels" could be misconstrued as indicating you're actually doing something to those pixels to actually calibrate them. When in reality you're just finding and flagging them all. The "self calibration" statement just means you're using the data itself to do some kind of operation on itself. So maybe consider some rewording of this comment.  Oh, and I also really like to see the use of punctuation in multi-line comments, i.e. if lines 248 and 249 are two separate comments, as opposed to one multi-line, run-on sentence, then they need periods at the ends of the lines.  ;-)

hbushouse commented on jwst/pipeline/calwebb_spec2.py:
```diff
@@ -254,10 +268,7 @@ def process_exposure_product(
-
         calibrated = self.imprint_subtract(calibrated, members_by_type['imprint'])
-
-        # for each background image subtract an associated leak cal
```
Why remove this comment? I think it's useful to let people know the purpose of lines 272-274 and how they differ from what was done to the science exposure in line 271.

emolter replied:
I'm putting it back because you're right that its removal is extraneous to this PR, but IMO this comment is redundant with the code itself.  It's obvious that the code is looping over background files and doing the imprint subtraction on the background files

hbushouse reviewed (approved):
Code looks good; RTD doc pages look good; regtest results look good. I approve!

hbushouse commented:
@emolter Feel free to merge whenever you want.
